File: backend/api/src/main/scala/kpn/api/common/data/Element.scala
package kpn.api.common.data

import kpn.api.common.data.raw.Raw
import kpn.api.custom.Tag
import kpn.api.custom.Timestamp

trait Element extends Meta with Tagable {

  def id: Long

  def version: Long

  def timestamp: Timestamp

  def changeSetId: Long

  def tags: Seq[Tag]

  def isNode: Boolean = false

  def isWay: Boolean = false

  def isRelation: Boolean = false

  def toMeta: MetaData = MetaData(version, timestamp, changeSetId)

  def raw: Raw = {
    Raw(
      version,
      changeSetId,
      timestamp,
      tags
    )
  }
}

File: backend/api/src/main/java/kpn/api/common/data/Element.java
package kpn.api.common.data;

import kpn.api.common.data.raw.Raw;
import kpn.api.custom.Tag;
import kpn.api.custom.Timestamp;

import com.google.common.collect.ImmutableList;

public interface Element {

  Long id();

  Long version();

  Timestamp timestamp();

  Long changeSetId();

  ImmutableList<Tag> tags();

  default boolean isNode() {
    return false;
  }

  default boolean isWay() {
    return false;
  }

  default boolean isRelation() {
    return false;
  }

  default MetaData toMeta() {
    return new MetaData(version(), timestamp(), changeSetId());
  }

  default Raw raw() {
    return new Raw(
      version(),
      changeSetId(),
      timestamp(),
      tags()
    );
  }
}
